### DISCUSSION TITLE
How can i reference Dependencies of a .NET core project

### DISCUSSION THREAD

**Discussion by BarakDabach** [agracio/edge-js#231] [Q&A]
I have an electron app that needs to use a compiled `.net 6` dll.
this library have  multiple dependencies.
according to the docs :
`NOTE: #r and references: [ 'MyDll.dll' ] references only work when using .NET Framework 4.5
If you are using .NET Core and are using the .NET Core SDK and CLI, you must have a project.json file (specification [here](https://github.com/aspnet/Home/wiki/Project.json-file)) that specifies the dependencies for the application. This list of dependencies must also include the [Edge.js runtime package](https://www.nuget.org/packages/EdgeJs/) and, if you need to be able to dynamically compile your code, the package(s) for the compilers that you plan to use, like [Edge.js.CSharp](https://www.nuget.org/packages/Edge.js.CSharp/).`

i understand that `#r and references: [ 'MyDll.dll' ` can only be used in `.net framwork`.
but didnt understand how in `.net core`

any help will be great!

thanks in advance

**Comment by agracio** [comment]
You should use compiled dlls instead of relying on inline code, take a look at https://github.com/agracio/electron-edge-js-quick-start for reference.

**Reply by BarakDabach** [reply]
I am using a pre compiled dll.
What i did not configured is the `process.env.EDGE_APP_ROOT` to the path of my dll bin folder.
Is that the reason why  edge did not properly resolve the dll dependendies?

**Reply by agracio** [reply]
Did you follow the same steps as in https://github.com/agracio/electron-edge-js-quick-start? You generally don't need to specify `process.env.EDGE_APP_ROOT` as long as your dll calls have correct paths. 
As I mentioned in previous reply you should not use inline c# code to try resolve references but follow the example to call dll methods.
You can fork https://github.com/agracio/electron-edge-js-quick-start and add calls to your dll there to make sure it works as intended.

**Reply by agracio** [reply]
Let me know if you need any help

**Comment by BarakDabach** [comment]
I've forked the repo  - will update

**Comment by BarakDabach** [comment]
In my real application i have pre compile legacy `dll` that talk to DB using the `Microsoft.Data.SqlClient` nuget package, version `5.2.2`

Test Description:
Forked from https://github.com/agracio/electron-edge-js-quick-start under branch  `sql-client-null-error`.
I've added to the QuickStart Solution under the `QuickStart.Core.LocalMethods` a method: `TestWithSqlDb()` -  that talks to a local DB(for demo purposes, just s simple `SELECT *`.
and in the electron app - a callback from edge-js to that method that displaying the result of the query.

Error:
When the callback excecute - i get a null exception from `.net` when trying to perform a fill on the select:
```
SqlDataAdapter select = new SqlDataAdapter("select * from The-Table-Name","Connection string");
DataSet dataSet = new DataSet(tableName);
select.Fill(dataSet, tableName);
```
To test that the method works correctly  - ive created a `Console App` - `TestWithDbRunner` that excecute the method:
everything works fine when calling the methods from `.net`.

the problem comes when calling it from electron.


scrennshot of the error:
![image](https://github.com/user-attachments/assets/50db102c-5808-44ff-affc-ffd77718d07d)

in order to recreate the error - need to setup a local DB.

hope it clear and detailed.
thank for the support

**Comment by agracio** [comment]
Can you show the full method that calls Fill, the one that you call using edge-js.

**Reply by BarakDabach** [reply]
```
public async Task<object> TestWithSqlDb(dynamic input)
        {
            var tableName = "adapters";
            SqlDataAdapter select = new SqlDataAdapter("select * from tableName",
                "Connection string to db");
            DataSet dataSet = new DataSet(tableName);
            select.Fill(dataSet, tableName);
            var adaptersTable = dataSet.Tables[0];
            var sourcesNames = new List<string>();
            foreach (DataRow a_row in adaptersTable.Rows)
            {
                
                sourcesNames.Add(a_row["Name"].ToString());
            }
            return string.Join(",",sourcesNames);
        }
```

**Reply by BarakDabach** [reply]
If you need more info - you can clone https://github.com/BarakDabach/electron-edge-js-quick-start.git under branch `sql-client-null-error`

**Reply by agracio** [reply]
Sorry did not read that it is in a branch, will take a look at the full code and run it.

**Comment by agracio** [comment]
Unfortunately unable to resolve your issue, System.NullReferenceException is thrown by SqlConnection object. Here is a sample code to reproduce:
```cs
public async Task<object> TestWithSqlDb(dynamic input)
{
    SqlConnection connection = new SqlConnection("Connection string to db");
    await connection.OpenAsync();
    return "";
}
```
No idea what is causing that.

**Reply by BarakDabach** [reply]
I have a hunch that maybe the issue is coming from electron security policies.
I will continue to investigate that, and will update,
@agracio  -  Thank you for the support  and maintaining this awesome library! 💪💪

**Reply by agracio** [reply]
It is not from Electron, I tested it on https://github.com/agracio/edge-js-quick-start to have a clearer view of the errors. 
It is likely something to do with dynamic loading of referenced dlls with some having issues.

**Comment by agracio** [comment]
In general both `Microsoft.Data.SqlClient` and `System.Data.SqlClient` have issues when loaded dynamically. When researching related problems saw issues posted related to Kubernetes loaders and some other dynamic loaders.

If you do not require SQL execution through your own dll you can use `edge-sql` instead. However it does not use `DataSet` object so any object manipulation would have to be done in resulting JSON. https://github.com/agracio/edge-sql

**Reply by BarakDabach** [reply]
Can you include links to the issues you found?

**Reply by agracio** [reply]
Unfortunately it was a long time ago when I was troubleshooting `edge-sql`. When working on this issue I had a feeling that it is not the first time I see this problem but took me a while to remember.

**Comment by agracio** [comment]
Narrowed it down to be the same issues as `System.Data.SqlClient` although they throw different exceptions.
For both only published applications work. 
For your quick-start example publish it using `dotnet publish src/QuickStart.sln --self-contained -r win-x64 -c Release` and change `baseNetAppPath` variable to `const baseNetAppPath = path.join(__dirname, '/src/QuickStart.Core/bin/Release/net8.0/win-x64/publish');`

This depends on a new `edge-js` and `electron-edge-js` release so should be done by tomorrow.

**Reply by BarakDabach** [reply]
I guess when not `self contained`,`.net` is unable to resolve the `SqlClient`

**Reply by agracio** [reply]
It resolves both dlls to a correct location in nuget packages but in case of Microsoft.Data.SqlClient does not load it's dependencies that are not specified in dependency manifest.

**Comment by agracio** [comment]
New `edge-js` and `electron-edge-js` releases complete. Unfortunately publishing your .NET app is the only solution for both SQL libs mentioned above.

**Reply by BarakDabach** [reply]
So i need to open a new pr right?

**Reply by agracio** [reply]
New PR, not sure what you mean. 
As per my comment above you can use your dll and can also test it with forked quick-start but you need to publish your .NET project first for Microsoft.Data.SqlClient and System.Data.SqlClient to work.

> For your quick-start example publish it using `dotnet publish src/QuickStart.sln --self-contained -r win-x64 -c Release` and change baseNetAppPath variable to `const baseNetAppPath = path.join(__dirname, '/src/QuickStart.Core/bin/Release/net8.0/win-x64/publish');`

Just make sure to update to latest version of `electron-edge-js` first.
